diff --git a/ac.scm b/ac.scm
index b16531f..7ce7754 100644
--- a/ac.scm
+++ b/ac.scm
@@ -772,17 +772,44 @@
                                 (current-output-port)))
                 b))
 
+(define (print* f port x)
+
+  (define (print x)
+    (cond ((pair? x)
+           (display "(" port)
+           (print (car x))
+           (print-cdr (cdr x)))
+          (#t
+           (f x port))))
+
+  (define (print-cdr y)
+    (cond ((or (eq? y 'nil) (eq? y '()))
+           (display ")" port))
+          ((pair? y)
+           (display " " port)
+           (print (car y))
+           (print-cdr (cdr y)))
+          (#t
+           (display " . " port)
+           (print y)
+           (display ")" port))))
+
+  (print x))
+
+(define (write*   x port) (print* write   port x))
+(define (display* x port) (print* display port x))
+
 (define (printwith f args)
   (let ((port (if (> (length args) 1)
                   (cadr args)
                   (current-output-port))))
     (when (pair? args)
-      (f (ac-denil (car args)) port))
+      (f (car args) port))
     (flush-output port))
     'nil)
 
-(xdef 'write (lambda args (printwith write   args)))
-(xdef 'disp  (lambda args (printwith display args)))
+(xdef 'write (lambda args (printwith write*   args)))
+(xdef 'disp  (lambda args (printwith display* args)))
 
 ; sread = scheme read. eventually replace by writing read
 
@@ -951,7 +978,7 @@
         (if (eqv? expr ':a)
             'done
             (let ((val (arc-eval expr)))
-              (write (ac-denil val))
+              (write* val (current-output-port))
               (namespace-set-variable-value! '_that val)
               (namespace-set-variable-value! '_thatexpr expr)
               (newline)
