diff --git a/srv.arc b/srv.arc
index f3bae12..efd13df 100644
--- a/srv.arc
+++ b/srv.arc
@@ -156,6 +156,8 @@ Connection: close"))
      '((gif       "image/gif")
        (jpg       "image/jpeg")
        (png       "image/png")
+       (css       "text/css")
+       (plain     "text/plain")
        (text/html "text/html; charset=utf-8")))
 
 (= rdheader* "HTTP/1.0 302 Moved")
@@ -238,8 +240,8 @@ Connection: close"))
            "jpg"  'jpg
            "jpeg" 'jpg
            "png"  'png
-           "css"  'text/html
-           "txt"  'text/html
+           "css"  'css
+           "txt"  'plain
            "htm"  'text/html
            "html" 'text/html
            "arc"  'text/html
